[webOS] Add new webOS videoplayer

sundermann opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
Adds a new videoplayer for webOS in favor of the old separate video codec and audio sink. Video codec and audio sink code are mostly merged into one pipeline and the code is roughly similar.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
With the current videoplayer where audio and video are handled separately, sync has always been an issue. This is because the public APIs we can use are very limited in that regard. Efforts to access the underlying gstreamer elements to get more advanced acess into synchronization were only partly successful.

That's why this PR introduces a new video player. It implements `IDVDStreamPlayerVideo` and `IDVDStreamPlayerAudio` feeding into a common pipeline. Synchronization is then handled by the webOS video APIs. It is still possible to set audio delay, subtitle delay etc.

There are a couple of commits improving the encoder and resampler. Those have been split into separate commits and should not affect platforms other than webOS.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
webOS 9
Users in the forum and on discord reported that sync issues are fixed and functionality on webOS 4 was restored.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
Fixes #24193
Fixes #25589
Fixes #25065
Potentially fixes #25101
Potentially fixes #25231

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [x] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

sundermann committed:
[webOS] New video player

sundermann committed:
[depends][webOS] Update webos-userland

sundermann committed:
[ffmpeg] Encoder: Smarter audio layout selection

Existing code does not pass a layout so it should not be affected by this change

sundermann committed:
[ffmpeg] Resampler: Create resampler if number of frames does not match

sundermann commented on xbmc/cores/VideoPlayer/MediaPipelineWebOS.cpp:
```diff
@@ -0,0 +1,1268 @@
+  return m_videoInfo;
+}
+
+bool CMediaPipelineWebOS::Supports(const AVCodecID codec)
```
Function is static.

sundermann commented on xbmc/cores/VideoPlayer/VideoPlayerWebOS.cpp:
```diff
@@ -0,0 +1,69 @@
+  {
+    bool hasAudio = m_SelectionStreams.CountType(STREAM_AUDIO);
+
+    delete m_VideoPlayerVideo;
+    delete m_VideoPlayerAudio;
+    m_mediaPipelineWebOS = std::make_unique<CMediaPipelineWebOS>(
+        *m_processInfo, m_renderManager, m_clock, m_messenger, m_overlayContainer, hasAudio);
+    m_VideoPlayerVideo = new CVideoPlayerVideoWebOS(*m_mediaPipelineWebOS, *m_processInfo);
+    m_VideoPlayerAudio = new CVideoPlayerAudioWebOS(*m_mediaPipelineWebOS, *m_processInfo);
+  }
+  else if (m_mediaPipelineWebOS || (!m_VideoPlayerVideo && !m_VideoPlayerAudio))
+  {
+    delete m_VideoPlayerVideo;
+    delete m_VideoPlayerAudio;
+    m_mediaPipelineWebOS = nullptr;
+    m_VideoPlayerVideo =
+        new CVideoPlayerVideo(&m_clock, &m_overlayContainer, m_messenger, m_renderManager,
+                              *m_processInfo, m_messageQueueTimeSize);
+    m_VideoPlayerAudio =
+        new CVideoPlayerAudio(&m_clock, m_messenger, *m_processInfo, m_messageQueueTimeSize);
```
Existing code. `Videoplayer` generally could make use of smart pointers but I don't want to make this PR bigger than it already is.

neo1973 commented on xbmc/cores/VideoPlayer/MediaPipelineWebOS.cpp:
```diff
@@ -0,0 +1,1278 @@
+    time["position"] = pts.count();
+    std::string payload;
+    CJSONVariantWriter::Write(time, payload, true);
+    if (!m_mediaAPIs->setTimeToDecode(payload.c_str()))
+    {
+      CLog::LogF(LOGERROR, "setTimeToDecode failed");
+      auto player = static_cast<mediapipeline::CustomPlayer*>(m_mediaAPIs->player.get());
+      auto pipeline = static_cast<mediapipeline::CustomPipeline*>(player->getPipeline().get());
+      MEDIA_CUSTOM_CONTENT_INFO_T contentInfo;
+      pipeline->loadSpi_getInfo(&contentInfo);
+      contentInfo.ptsToDecode = pts.count();
+      pipeline->setContentInfo(MEDIA_CUSTOM_SRC_TYPE_ES, &contentInfo);
+    }
+
+    auto player = static_cast<mediapipeline::CustomPlayer*>(m_mediaAPIs->player.get());
+    auto pipeline = static_cast<mediapipeline::CustomPipeline*>(player->getPipeline().get());
```
```suggestion
    auto player = static_cast<mediapipeline::CustomPlayer*>(m_mediaAPIs->player.get());
    auto pipeline = static_cast<mediapipeline::CustomPipeline*>(player->getPipeline().get());
    
    if (!m_mediaAPIs->setTimeToDecode(payload.c_str()))
    {
      CLog::LogF(LOGERROR, "setTimeToDecode failed");
      MEDIA_CUSTOM_CONTENT_INFO_T contentInfo;
      pipeline->loadSpi_getInfo(&contentInfo);
      contentInfo.ptsToDecode = pts.count();
      pipeline->setContentInfo(MEDIA_CUSTOM_SRC_TYPE_ES, &contentInfo);
    }
```

neo1973 commented on xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp:
```diff
@@ -36,9 +36,10 @@ CAEEncoderFFmpeg::CAEEncoderFFmpeg() : m_CodecCtx(NULL), m_SwrCtx(NULL)
+  if (m_SwrCtx)
+    swr_free(&m_SwrCtx);
+  if (m_CodecCtx)
+    avcodec_free_context(&m_CodecCtx);
```
`swr_free` and `avcodec_free_context` check themself if they are already null.

https://github.com/FFmpeg/FFmpeg/blob/722a2170e83231283fc74bede495b3b4ee9591ac/libswresample/swresample.c#L121

https://github.com/FFmpeg/FFmpeg/blob/722a2170e83231283fc74bede495b3b4ee9591ac/libavcodec/options.c#L164

sundermann replied:
Interesting, I've removed the checks. `av_channel_layout_uninit` was causing the issue as the access to the member variable is unsafe as `m_CodecCtx` can be null. I've changed the commit. The whole encoder class could use an overhaul. Half of the code is completely unused (like m_SwrCtx which is set but never used).

neo1973 reviewed (commented):
Sounds very exciting! Unfortunately I don't have the hardware to test, but browsing through the code I spotted a couple of places that may need attention.

sundermann committed:
[ffmpeg] Encoder: Do not free m_CodecCtx->chLayout

It is already being freed as part of avcodec_free_context, and it can cause a segfault if m_CodecCtx is null

sundermann commented on xbmc/cores/VideoPlayer/MediaPipelineWebOS.cpp:
```diff
@@ -0,0 +1,1278 @@
+    s << ", transcoded ac3";
+  }
+
+  m_audioInfo = s.str();
```
Does moving the `std::ostringstream` do anything?

neo1973 replied:
It causes the rvalue overload of [`std::ostringstream::str()`](https://en.cppreference.com/w/cpp/io/basic_ostringstream/str.html) to be called which moves the underlying `std::string` out of the stream instead of copying it.

neo1973 replied:
> I think `std::ostringstream` doesn't have a move-optimized version.

Overload (3) from the link?

sundermann replied:
Thanks for the explaining! I didn't knew there was a new overload in C++20. Commited the change

sundermann replied:
I actually refactored these functions to use `std::format` which is a lot more readable (and faster?) than the `stringstream` approach.

garbear commented on xbmc/cores/VideoPlayer/MediaPipelineWebOS.h:
```diff
@@ -0,0 +1,440 @@
+ * @class CVideoPlayerAudioWebOS
+ * @brief Audio stream player adapter forwarding to CMediaPipelineWebOS.
+ */
+class CVideoPlayerAudioWebOS final : public IDVDStreamPlayerAudio
```
Can separate classes be split into separate files? I got a little scared when I saw a 440 line header, not realizing it was multiple classes.

sundermann replied:
Sure but those two classes mostly just forward into the pipeline class.

garbear replied:
OK.

sundermann replied:
They live in their own header files now. Do you think that's alright or should I also split the definitions into their own cpp files?

garbear replied:
I'd split. Files are cheap, and long headers should be avoided. IMO it also helps to see the architecture at a FS level before diving into the code.

sundermann replied:
Done, CLions split function definition is great

garbear commented on xbmc/cores/VideoPlayer/MediaPipelineWebOS.cpp:
```diff
@@ -0,0 +1,1275 @@
+/*
+*  Copyright (C) 2005-2025 Team Kodi
```
Oopsie

garbear reviewed (approved):
I didn't scan MediaPipelineWebOS.cpp in depth, but at a level above that this looks good. I like the documentation in CMediaPipelineWebOS.

sundermann commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererStarfish.cpp:
```diff
@@ -108,26 +109,29 @@ void CRendererStarfish::ManageRenderArea()
+          static_cast<long>(m_sourceRect.y1), static_cast<long>(m_sourceRect.Width()),
+          static_cast<long>(m_sourceRect.Height()), static_cast<long>(m_destRect.x1),
+          static_cast<long>(m_destRect.y1), static_cast<long>(m_destRect.Width()),
+          static_cast<long>(m_destRect.Height()), false, &m_videoBuffer->m_taskId);
```
It can't be null

sundermann commented on xbmc/cores/VideoPlayer/MediaPipelineWebOS.cpp:
```diff
@@ -0,0 +1,1413 @@
+  if (result.find("BufferFull") != std::string::npos)
+  {
+    m_messageQueueAudio.PutBack(msg);
+    std::this_thread::sleep_for(100ms);
```
Intentional

garbear commented on xbmc/cores/VideoPlayer/MediaPipelineWebOS.h:
```diff
@@ -0,0 +1,385 @@
+{
+public:
+  /**
+  * @brief Construct the WebOS media pipeline.
```
Still bad indentation in this file.

garbear commented on xbmc/cores/VideoPlayer/MediaPipelineWebOS.h:
```diff
@@ -37,9 +37,9 @@ class CDVDAudioCodec;
-  * @class CMediaPipelineWebOS
-  * @brief WebOS media pipeline for audio/video playback.
-  */
+* @class CMediaPipelineWebOS
```
Your IDE just does not like you.

sundermann replied:
Looks alright to me? I don't know what the exact formatting rules on doxygen comments are. I looked at a couple of files for reference but it's all over the place and clang-format generally doesn't format these at all

sundermann replied:
I just found `ReflowComments: IndentOnly` which does allow clang-format to properly work. Should be alright now. Maybe we should set that clang-format option by default though.

sundermann committed:
[MessageQueue] Add missing lock in IsDataBased

IsDataBased is part of the public API. It uses the members m_TimeBack and m_TimeFront which can be modified by another thread. For the calls from within this class it's properly locked but external callers also need to acquire the lock. m_section is a recursive lock so adding the lock here is safe.

sundermann commented:
@rcorex We've had people on the forums with a C9 and they've not reported that the screensaver wasn't working. Weird stuff. Unfortunately, webOS 4 is vastly different from webOS 5+ and I do not own one myself.

I want to move forward with this PR though and hope to figure out why that's happening (sometimes?) on webOS 4 and address it in a followup PR.

jeanmoulart commented:
@sundermann I was about to get a shield pro because of the audio sync issue I couldn't fix with any of the suggested workarounds. This commit solved all audio issues I had encountered.

Thanks a **lot**!
